File: python_client/src/ezauth_client/rows.py
from __future__ import annotations

from typing import TYPE_CHECKING

from ._client import _encode

if TYPE_CHECKING:
    from ._client import BaseClient


class Rows:
    def __init__(self, client: BaseClient):
        self._client = client

    def insert(self, table_id: str, data: dict, *, user_id: str | None = None) -> dict:
        body: dict = {"data": data}
        if user_id is not None:
            body["user_id"] = user_id
        return self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/rows",
            method="POST",
            body=body,
            auth="auto",
        )

    def get(self, table_id: str, row_id: str) -> dict:
        return self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/rows/{_encode(row_id)}",
            auth="auto",
        )

    def update(self, table_id: str, row_id: str, data: dict) -> dict:
        return self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/rows/{_encode(row_id)}",
            method="PATCH",
            body={"data": data},
            auth="auto",
        )

    def delete(self, table_id: str, row_id: str) -> None:
        self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/rows/{_encode(row_id)}",
            method="DELETE",
            auth="auto",
        )

    def query(
        self,
        table_id: str,
        *,
        filter: dict | None = None,
        sort: dict | list | None = None,
        limit: int | None = None,
        cursor: str | None = None,
    ) -> dict:
        body: dict = {}
        if filter is not None:
            body["filter"] = filter
        if sort is not None:
            body["sort"] = sort
        if limit is not None:
            body["limit"] = limit
        if cursor is not None:
            body["cursor"] = cursor
        return self._client._fetch(
            f"/v1/tables/{_encode(table_id)}/rows/query",
            method="POST",
            body=body,
            auth="auto",
        )

File: javascript_client/src/rows.js
export class Rows {
  constructor(client) {
    this._client = client;
  }

  insert(tableId, { data, userId } = {}) {
    const body = { data };
    if (userId !== undefined) body.user_id = userId;
    return this._client._fetch(`/v1/tables/${encodeURIComponent(tableId)}/rows`, {
      method: 'POST',
      body,
      auth: 'auto',
    });
  }

  get(tableId, rowId) {
    return this._client._fetch(
      `/v1/tables/${encodeURIComponent(tableId)}/rows/${encodeURIComponent(rowId)}`,
      { auth: 'auto' },
    );
  }

  update(tableId, rowId, { data }) {
    return this._client._fetch(
      `/v1/tables/${encodeURIComponent(tableId)}/rows/${encodeURIComponent(rowId)}`,
      { method: 'PATCH', body: { data }, auth: 'auto' },
    );
  }

  delete(tableId, rowId) {
    return this._client._fetch(
      `/v1/tables/${encodeURIComponent(tableId)}/rows/${encodeURIComponent(rowId)}`,
      { method: 'DELETE', auth: 'auto' },
    );
  }

  query(tableId, { filter, sort, limit, cursor } = {}) {
    const body = {};
    if (filter !== undefined) body.filter = filter;
    if (sort !== undefined) body.sort = sort;
    if (limit !== undefined) body.limit = limit;
    if (cursor !== undefined) body.cursor = cursor;

    return this._client._fetch(
      `/v1/tables/${encodeURIComponent(tableId)}/rows/query`,
      { method: 'POST', body, auth: 'auto' },
    );
  }
}
